frontend/lib/src/dataframes/arrowFormatUtils.ts

```diff
@@ -144,89 +151,147 @@ interface Interval {
+  } else if (unit === TimeUnit.NANOSECOND) {
     // Nanoseconds
     unitAdjustment = 1000 * 1000 * 1000
   } else {
     // Interpret it as seconds as a fallback
-    return Number(value)
+    return Number(timestamp)
   }
 
   // Do the calculation based on bigints, if the value
   // is a bigint and not safe for usage as number.
   // This might lose some precision since it doesn't keep
   // fractional parts.
-  if (typeof value === "bigint" && !Number.isSafeInteger(Number(value))) {
-    return Number(value / BigInt(unitAdjustment))
+  if (
+    typeof timestamp === "bigint" &&
+    !Number.isSafeInteger(Number(timestamp))
+  ) {
+    return Number(timestamp / BigInt(unitAdjustment))
   }
 
-  return Number(value) / unitAdjustment
+  return Number(timestamp) / unitAdjustment
 }
 
-function formatTime(data: number, field?: Field): string {
-  const timeInSeconds = convertToSeconds(data, field?.type?.unit ?? 0)
-  return moment
-    .unix(timeInSeconds)
+/**
+ * Converts a timestamp to a date object.
+ *
+ * @param timestamp The timestamp to convert.
+ * @param field The field containing the unit information.
+ * @returns The date object.
+ */
+export function convertTimestampToDate(
+  timestamp: number | bigint,
+  field?: Field
+): Date {
+  // Time values from arrow are not converted to a shared unit and
+  // just return the raw arrow value. Therefore, we need to adjust
+  // the value to seconds based on the unit information in the field.
+  // https://github.com/apache/arrow/blob/9e08c57c0986531879aadf7942998d26a94a5d1b/js/src/visitor/get.ts#L193C7-L209
+  const timeInSeconds = convertTimestampToSeconds(
+    timestamp,
+    field?.type?.unit ?? TimeUnit.SECOND
+  )
+  return moment.unix(timeInSeconds).utc().toDate()
+}
+
+/**
+ * Formats a time value based on the unit information in the field.
+ *
+ * @param timestamp The time value to format.
+ * @param field The field containing the unit information.
+ * @returns The formatted time value.
+ */
+function formatTime(timestamp: number | bigint, field?: Field): string {
+  const date = convertTimestampToDate(timestamp, field)
+  return moment(date)
     .utc()
-    .format(timeInSeconds % 1 === 0 ? "HH:mm:ss" : "HH:mm:ss.SSS")
+    .format(date.getMilliseconds() === 0 ? "HH:mm:ss" : "HH:mm:ss.SSS")
 }
 
-function formatDuration(data: number | bigint, field?: Field): string {
+function formatDate(date: number | Date): string {
+  const formatPattern = "YYYY-MM-DD"
+
+  // Date values from arrow are already converted to a date object
+  // or a timestamp in milliseconds even if the field unit might indicate a
+  // different unit.
+  // https://github.com/apache/arrow/blob/9e08c57c0986531879aadf7942998d26a94a5d1b/js/src/visitor/get.ts#L167-L171
+  if (
+    date instanceof Date ||
+    (typeof date === "number" && Number.isFinite(date))
+  ) {
+    return moment.utc(date).format(formatPattern)
+  }
+
+  logWarning(`Unsupported date type: ${date}`)
+  return String(date)
```

Comment:
Should we raise an error instead? It could be confusing to get a non-formatted string back here and this should be made more explicit. What do you think?

Reply:
The format function is mostly a best-effort logic to make the value look nicer for humans. It's not too big of a deal if some niche situations do not look as nice. Raising exception would make it a lot harder to use this logic in the consumers (dataframe/table) since they would have to somehow deal with the exception on a cell level to not crash the full table or app. So, I think it is good enough to just log a warning and convert it to string if we are unsure on how to deal with it. If formatting for important values is broken, I expect it to eventually bubble up to us through some GH or internal issues.

Reply:
Gotcha, that makes sense to me. Could you capture this reasoning in a comment here, please? 🙂

Reply:
I added a comment about not throwing exceptions to the higher-level `format` function since it is relevant for all the underlying format functions